src/components/DnsSettings.tsx

```diff
@@ -1,313 +1,239 @@
-    protocol: this.props.settings!.dns?.protocol || DnsProtocol.UDP,
-    port: this.props.settings!.dns?.port || 53,
-
-    // validation
-    validRCodes: this.props.settings!.dns?.validRCodes || [],
-    validateAnswerRRS: this.props.settings!.dns?.validateAnswerRRS,
-    validateAuthorityRRS: this.props.settings!.dns?.validateAuthorityRRS,
-    validateAdditionalRRS: this.props.settings!.dns?.validateAdditionalRRS,
-  };
-
-  onUpdate = () => {
-    const settings = this.state as DnsSettings;
-    this.props.onUpdate({
-      dns: settings,
-    });
-  };
-
-  onIpVersionChange = (value: SelectableValue<IpVersion>) => {
-    this.setState({ ipVersion: value.value || IpVersion.Any }, this.onUpdate);
-  };
-
-  onProtocolChange = (value: SelectableValue<DnsProtocol>) => {
-    this.setState({ protocol: value.value || DnsProtocol.UDP }, this.onUpdate);
-  };
-
-  onPortChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-    this.setState({ port: event.target.valueAsNumber || 53 }, this.onUpdate);
-  };
-
-  onServerChange = (event: React.ChangeEvent<HTMLInputElement>) => {
-    this.setState({ server: event.target.value }, this.onUpdate);
-  };
-
-  onRecordTypeChange = (value: SelectableValue<DnsRecordType>) => {
-    this.setState({ recordType: value.value || DnsRecordType.A }, this.onUpdate);
-  };
-
-  onShowAdvanced = (isOpen: boolean) => {
-    this.setState({ showAdvanced: !this.state.showAdvanced });
-  };
-
-  onShowValidation = (isOpen: boolean) => {
-    this.setState({ showValidation: !this.state.showValidation });
-  };
-
-  onValidRCodesChange = (item: Array<SelectableValue<string>>) => {
-    let validRCodes: string[] = [];
-    for (const p of item.values()) {
-      if (p.value) {
-        validRCodes.push(p.value);
-      }
-    }
-    this.setState({ validRCodes }, this.onUpdate);
-  };
-
-  onValidateAnswerChange = (validations: DNSRRValidator | undefined) => {
-    if (!validations) {
-      this.setState({ validateAnswerRRS: undefined }, this.onUpdate);
-      return;
-    }
-    const validateAnswerRRS = {
-      failIfMatchesRegexp: validations.failIfMatchesRegexp,
-      failIfNotMatchesRegexp: validations.failIfNotMatchesRegexp,
-    };
-    this.setState({ validateAnswerRRS }, this.onUpdate);
+function dnsSettingsReducer(state: DnsSettings, action: Action) {
+  return {
+    ...state,
+    [action.name]: action.value ?? action.fallbackValue,
   };
+}
 
-  onValidateAuthorityChange = (validations: DNSRRValidator | undefined) => {
-    if (!validations) {
-      this.setState({ validateAuthorityRRS: undefined }, this.onUpdate);
-      return;
-    }
-    const validateAuthorityRRS = {
-      failIfMatchesRegexp: validations.failIfMatchesRegexp,
-      failIfNotMatchesRegexp: validations.failIfNotMatchesRegexp,
-    };
-    this.setState({ validateAuthorityRRS }, this.onUpdate);
-  };
+const DnsSettingsForm: FC<Props> = ({ settings, isEditor, onUpdate }) => {
+  const initialValues = {
```

Comment:
Wonder if we could defined a default object and merge it with `settings.dns` instead?
```
const defaultValues = {
  recordType: DnsRecordType.A,
  server: '8.8.8.8'
  //.... 
}
const initialValues = {...defaultValues, ...settings.dns};
```